rfcs/0007-handling-corruption.md

```diff
@@ -0,0 +1,533 @@
+# Handling Data Corruption
+
+Status: Draft
+
+Authors:
+
+* [Derrick J Wippler](https://github.com/thrawn01)
+
+References:
+
+- https://www.sqlite.org/atomiccommit.html
+- https://avi.im/blag/2024/sqlite-bit-flip/
+- https://github.com/danthegoodman1/BreakingSQLite
+- https://www.sqlite.org/recovery.html
+- https://www.postgresql.org/docs/9.1/app-pgresetxlog.html
+- https://www.postgresql.org/docs/current/app-initdb.html#APP-INITDB-DATA-CHECKSUMS
+- https://www.postgresql.org/docs/current/runtime-config-developer.html#GUC-IGNORE-CHECKSUM-FAILURE
+- https://www.postgresql.org/docs/current/runtime-config-developer.html#GUC-ZERO-DAMAGED-PAGES
+- https://docs.tigerbeetle.com/about/safety/
+- https://github.com/thrawn01/db-corruption
+
+## Motivation
+The durability of object storage is a significant factor driving the adoption
+of disaggregated storage. However, we should not assume that data corruption
+will never occur. Several factors can lead to corruption, including:
+
+- Programming errors in SlateDB
+- Malicious intent to corrupt data
+- Variations in the logical durability of different object store implementations
+- Errors in firmware and drivers
+
+How SlateDB operates in the face of corruption is crucial for its adoption and
+for maintaining operator confidence.
+
+Although SlateDB utilizes object stores that typically provide some level of
+Logical Availability and Durability, SlateDB should be viewed as a
+manifestation of a physical system. This is because SlateDB cannot make
+assumptions about the Logical Availability or Durability of the object store it
+uses to store data. The object store provided to SlateDB at startup could
+simply be a local disk masquerading as a highly available logical store.
+
+### Prioritizing Durability
+Prioritizing durability means that in the event of data corruption, the
+database will immediately become unavailable until the corruption is repaired
+through some form of out-of-band intervention. This approach ensures that the
+database cannot lose data without the operator's knowledge.
+
+### Prioritizing Availability
+Prioritizing availability means that the database will acknowledge the
+corruption, attempt to recover as much data as possible, and warn the operator,
```

Comment:
Point of clarification: "as much data as possible" -- Is this what you mean to say. To me, that implies that it could run heavy weight/slow recovery processes. I suspect what you mean to say is that the database will skip (ignore) corrupt data, which IMO is slightly different.